Feedback ctrl

garygra opened the pull request:
Feedback control setup:
  * Adding srv and msgs dependencies
  * Adding sensors to mushr
  * State msg definition
  * Feedback service definition
  * Feedback service sim
  * Reading from the sensors to define the state
  * Adding scripts into Cmake
  * Global frequency variable to handle mujoco/ros timesteps
  * feedback service class
  * simulator uses ros::rate for stepping.
  * Sim viz uses ros::WallRate to always update even if using sim_time
  * keyboard service to *simulate* a feedback controller
  * mushr_feedback implementation
  * launch feedback example

garygra committed:
Feedback control setup:
  * Adding srv and msgs dependencies
  * Adding sensors to mushr
  * State msg definition
  * Feedback service definition
  * Feedback service sim
  * Reading from the sensors to define the state
  * Adding scripts into Cmake
  * Global frequency variable to handle mujoco/ros timesteps
  * feedback service class
  * simulator uses ros::rate for stepping.
  * Sim viz uses ros::WallRate to always update even if using sim_time
  * keyboard service to *simulate* a feedback controller
  * mushr_feedback implementation
  * launch feedback example

aravindsiv committed:
Merge branch 'master' into feedback-ctrl

aravindsiv commented on mj_models/CMakeLists.txt:
```diff
@@ -8,7 +8,7 @@ set(CMAKE_CXX_STANDARD 17)
 find_package(catkin REQUIRED COMPONENTS
   roscpp
-  std_msgs
+  std_msgs geometry_msgs
```
Nit: New line for readability

aravindsiv commented on mj_models/CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.0.2)
+cmake_minimum_required(VERSION 3.5)
```
Is there any specific reason we need `3.5` instead of maybe `3.2`? I

garygra replied:
It might be only my machine, but I get:
```
CMake Deprecation Warning at mujoco_ros/mujoco_ros/CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.
```

aravindsiv commented on mj_models/msg/mushr/MushrState.msg:
```diff
@@ -0,0 +1,4 @@
+std_msgs/Header header
+geometry_msgs/Pose pose
+std_msgs/String[] str_extra # For additional non-pose data 
```
Why do we have this?

aravindsiv commented on mj_models/msg/mushr/MushrState.msg:
It would be more appropriate to name this something like `MushrObservation` or `MushrObservedState` or `MushrEstimatedState`.

garygra replied:
I like MushrObservation

aravindsiv commented on mujoco_ros/include/mujoco_ros/defs.h:
```diff
@@ -6,4 +8,13 @@
+{
+namespace constants
+{
+static double frequency{ 50.0 };
```
As we discussed offline, maybe this should not be a constant and set instead as `1.0/m->opt.timestep`.

aravindsiv commented on mujoco_ros/include/mujoco_ros/simulator.hpp:
```diff
@@ -18,6 +17,8 @@ class mujoco_simulator_t : public std::enable_shared_from_this<mujoco_simulator_
 
   GLFWwindow* window;
 
+  double _frequency;
```
We can remove this, right?

aravindsiv commented on mj_models/include/mj_models/mj_mushr.hpp:
```diff
@@ -1,15 +1,28 @@
 struct mushr_t
 {
   static constexpr std::size_t u_dim{ 2 };
+  struct sensors_t
```
Maybe something more descriptive, like `ground_truth_pose_sensor`.

garygra replied:
I intended this to be for *any* sensor, not only pose (i.e. IMU)

aravindsiv commented on mj_models/include/mj_models/mj_mushr.hpp:
```diff
@@ -20,4 +33,16 @@ inline void copy(Ctrl ctrl_out, const mj_models::MushrControl& msg)
 }
 
+template <typename Sensor>
+inline void state_from_sensors(mj_models::MushrState& msg, const Sensor& sensordata)
```
See above -- since these will not be the sensors that we will be using for the real system (or even a simulated version in experiments), we should ideally make the distinction between the ground truth pose and what a sensor provides.

aravindsiv reviewed (changes requested):
I will check the feedback files logic, but in the meantime, there are some minor changes to address.

garygra committed:
Addressing PR comments

aravindsiv commented on mj_models/msg/mushr/MushrObservation.msg:
```diff
@@ -0,0 +1,4 @@
+std_msgs/Header header
+geometry_msgs/Pose pose
+std_msgs/String[] str_extra # For additional non-pose data 
```
What are these for?

aravindsiv commented on mujoco_ros/scripts/keyboard_service.py:
```diff
@@ -0,0 +1,80 @@
+steering_increase = 0.1
+
+def handle_service(req):
+    global control
```
I am not a fan of Python `global`. It makes sense to put all of these in a Python class. Let me work on that.

aravindsiv commented on mujoco_ros/launch/mushr_feedback_example.launch:
```diff
@@ -0,0 +1,10 @@
+        <node pkg="mujoco_ros" name="mushr_feedback" type="mushr_feedback" output="screen" >
+            <param name="visualize" value="true" />
+        </node>
+        <!-- <node pkg="mujoco_ros" name="keyboard_service" type="keyboard_service.py" output="screen" launch-prefix="xterm -e"/> -->
```
Why is this removed?

garygra replied:
In some setups, is easier to launch this in a separate terminal (and the launch-prefix for a separate terminal not always works)

aravindsiv reviewed (changes requested):
Only some minor comments that we can discuss offline, and then I will make a change to the keyboard service.

aravindsiv committed:
1. Make keyboard service more object-oriented
2. Remove `mushr_control_listener`
3. Use namespace while accessing topics

garygra committed:
1. Fixing reset service not working.
2. Fixed random crash due to reset: adding reset mutex.
3. Adding `run_simulation` function to encapsulate running sim, viz and callbacks
4. Added mj_ros namespace to simulator and fixed names removing "mujoco"
5. Made `simulator_t` constructor private to force using a shared_ptr always through `simulator_t::initialize()`.

aravindsiv committed:
Namespace fixes

aravindsiv reviewed (approved):
LGTM, we can merge.
